src/main/java/org/spongepowered/api/block/tile/Jukebox.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered.org <http://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.block.tile;
+
+import org.spongepowered.api.block.tile.data.JukeboxData;
+import org.spongepowered.api.item.inventory.ItemStack;
+
+/**
+ * Represents a jukebox, also know as a record player.
+ */
+public interface Jukebox extends TileEntity {
+
+    /**
+     * Attempts to play the currently stored record according to the
+     * {@link JukeboxData} of this {@link Jukebox}.
+     */
+    void playRecord();
+
+    /**
+     * Ejects the record item in this Jukebox into the world.
+     */
+    void ejectRecord();
+
+    /**
+     * Ejects the current record in this Jukebox and inserts the given one.
+     *
+     * @param record The record to insert
+     */
+    void insertRecord(ItemStack record);
+
+    /**
+     * Gets the associated {@link JukeboxData} for this {@link Jukebox}.
+     *
+     * @return The current associated juke box data
+     */
+    JukeboxData getJukeboxData();
+
+    /**
+     * Sets the given {@link JukeboxData} to this {@link Jukebox}.
+     *
+     * <p>While {@link JukeboxData} is meant for {@link Jukebox}es, the data
+     * is validated.</p>
+     *
+     * @param data The jukebox data to set
+     * @return The transaction result
+     */
+    TileDataTransactionResult setJukeboxData(JukeboxData data);
```

Comment:
What about my suggestion of a `fromData` method? `jukebox.getJukeboxData().fromData(itemStackData)`
It could be added to `TileEntityData` so all tile entities have support for it without defining a setter in every TE